examples/constraints.rs

```diff
@@ -0,0 +1,217 @@
+use std::{error::Error, io};
+
+use crossterm::{
+    event::{self, DisableMouseCapture, EnableMouseCapture, Event, KeyCode},
+    execute,
+    terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen},
+};
+use itertools::Itertools;
+use ratatui::{layout::Constraint::*, prelude::*, widgets::*};
+
+fn main() -> Result<(), Box<dyn Error>> {
+    // setup terminal
+    enable_raw_mode()?;
+    let mut stdout = io::stdout();
+    execute!(stdout, EnterAlternateScreen, EnableMouseCapture)?;
+    let backend = CrosstermBackend::new(stdout);
+    let mut terminal = Terminal::new(backend)?;
+
+    // create app and run it
+    let res = run_app(&mut terminal);
+
+    // restore terminal
+    disable_raw_mode()?;
+    execute!(
+        terminal.backend_mut(),
+        LeaveAlternateScreen,
+        DisableMouseCapture
+    )?;
+    terminal.show_cursor()?;
+
+    if let Err(err) = res {
+        println!("{err:?}");
+    }
+
+    Ok(())
+}
+
+fn run_app<B: Backend>(terminal: &mut Terminal<B>) -> io::Result<()> {
+    let mut selection = 0;
+    loop {
+        terminal.draw(|f| ui(f, selection))?;
+
+        if let Event::Key(key) = event::read()? {
+            use KeyCode::*;
+            match key.code {
+                Char('q') => break Ok(()),
+                Char('j') | Char('l') | Down | Right => {
+                    selection = selection.saturating_add(1).min(6)
+                }
+                Char('k') | Char('h') | Up | Left => selection = selection.saturating_sub(1),
+                _ => (),
+            }
+        }
+    }
+}
+
+struct Example {
+    constraints: Vec<Constraint>,
+}
+
+impl Example {
+    fn new<C>(constraints: C) -> Self
+    where
+        C: Into<Vec<Constraint>>,
+    {
+        Self {
+            constraints: constraints.into(),
+        }
+    }
+}
+
+impl Widget for Example {
+    fn render(self, area: Rect, buf: &mut Buffer) {
+        let [title, legend, area] = area.split(&Layout::vertical([Ratio(1, 3); 3]));
+        let heading = Line::from(
+            Itertools::intersperse(
+                self.constraints.iter().enumerate().map(|(i, c)| {
+                    Span::styled(
+                        format!("{:?}", c),
+                        Style::default().fg(Color::Indexed(i as u8 + 1)),
+                    )
```

Comment:
```suggestion
                    let color = Color::Indexed(i as u8 + 1);
                    Span::styled(format!("{:?}", c), color)
```
Also, wonder it it might be worth putting a small macro implementation of this (without the need to discuss style interpolations because spans only have a single style) so we could just have written:
```rust
span!("{:?}", c, color);
```

Reply:
I like the idea!

Reply:
#796